### ISSUE TITLE
`stdin` and `stdout` performance considerations are not documented

### ISSUE THREAD

**Issue by Shnatsel** [rust-lang/rust#106133]
### Location
 - https://doc.rust-lang.org/stable/std/io/struct.Stdout.html
 - https://doc.rust-lang.org/stable/std/io/fn.stdout.html
 - https://doc.rust-lang.org/stable/std/io/struct.Stdin.html
 - https://doc.rust-lang.org/stable/std/io/fn.stdin.html

### Summary
It is a common performance pitfall for newcomers to write to `stdin` or `stdout` like they would in other languages: 

```rust
io::stdout().write_all(b"hello world")?;
```

However, this acquires a lock on every access, and the naive version slow if such writes are done frequently. A faster version looks like this:

```rust
    let stdout = io::stdout();
    let mut handle = stdout.lock();
    handle.write_all(b"hello world")?;
```

All pages linked do provide this example code, but do not explain **why** this more verbose version is beneficial, instead simply titling it "Using explicit synchronization:". None of these pages mention performance at all.

The relation to buffering is also not documented. Do `stdin` and `stdout` need to be wrapped into `BufReader`? What about the explicitly locked version?

**Comment by the8472** [comment]
> A faster version looks like this:

This isn't any faster because the `write_all` method [is overridden](https://doc.rust-lang.org/stable/src/std/io/stdio.rs.html#709-711) and only takes the lock once for the whole string even when short writes occur.

Taking the lock only makes sense when user code makes more than one call.

**Comment by alexpyattaev** [comment]
The key issue with all of the BufRead related API is that just about every function will always allocate, for every string read. This is putting enormous pressure onto the allocator, which results in poor performance. The only way around this in today's Rust is to dive down into the implementation of the reader and access the raw bytes, which is over 10x faster in my measurements. For example, the code below will count lines about 100x faster than 'naive' implementation that is using Stdin.readline()

``` rust
fn countlines()->usize{ 
      let mut cnt:usize = 0;
      let mut stdin = io::stdin().lock();
      loop{
          let used = {
              let available = match stdin.fill_buf() {
              Ok(n) => n,          
              Err(e) => {panic!("AAA")},
              };
              if available.len() == 0 {
                  break;
              }
 
              for b in available{
                  cnt += (*b == ('\n' as u8)) as usize;
              }
              
              available.len()
          };
          stdin.consume(used);
      }
      cnt
  }
```

While this is still technically Rust, I'd argue it is not great that one has to dive to such low level to get performance that comes "for free" in C:
``` c++
size_t countlines(void)
  {
    int fd = 0;
    ssize_t bytes_read;
    char buf[BUFFER_SIZE + 1];
    size_t lines = 0;
  
    while ((bytes_read = read (fd, buf, BUFFER_SIZE)) > 0)
      {
        char *p = buf;
        char *end = buf + bytes_read;          
        while (p != end)
        {
          lines += *p++ == '\n';
        }
      }
 return size_t;
}

```

**Comment by hillu** [comment]
`stdin.read_line()` fills a `String`, so it has to do extra work to ensure that this String is valid UTF-8. Does `stdin.read_until()` perform as badly in your measurements? Would you like to share your "naive" test program?

**Comment by alexpyattaev** [comment]
Another question that are worth looking into
   - would it make sense to have a version of stdin that does *not* take locks by default, and therefore is not Send (i.e. you'd have to manually wrap it to send to other threads)? And then this version can be substituted instead of a fully lock-guarded one whenever program is suspected to be using threading. Not sure if that is actually feasible in rustc though.
   - would it be fair to assume that functions such as println! also implicitly take and release stdout locks all the time, even in a single threaded application? In this case, they could be sped up substantially as well.

**Comment by lolbinarycat** [comment]
@alexpyattaev such a function would have to be callable only once, or it could be called multiple times on multiple threads.

additionally, there's no way for rustc to know if a program is single-threaded, since it could be using a platform-specific threading api instead of the one in `std::thread`.

**Comment by lolbinarycat** [comment]
the rust benchmarks were a bit misleading, see https://github.com/alexpyattaev/stdin_torture/pull/1

**Comment by alexpyattaev** [comment]
I guess this is primarily a matter of documentation. Given the rust memory model, I doubt this can be made substantially faster.

**Comment by lolbinarycat** [comment]
if there's any code here that can be improved, it's the zero-contention overhead of the locks, but rust just uses the OS locks instead of rolling its own.

**Comment by lolbinarycat** [comment]
i do think `lines` could use a performance disclaimer/section, since in simple programs the allocation per line may be more expensive than the entirety of their application logic.

**Comment by alexpyattaev** [comment]
I guess the best thing would be an API that locks the stream, loads data into an externally allocated buffer, and then borrows from that buffer via iterator over lines. Granted, this could fail miserably if the buffer is smaller than the longest line, but it would be rather fast.

**Comment by Shnatsel** [comment]
I've spent some time exploring the fastest ways to read a file in Rust and contributed my findings to the Rust Performance Book, see [their section on I/O](https://nnethercote.github.io/perf-book/io.html).

However, this issue is purely about documentation for locking overhead. Allocations and other advanced use cases are out of scope for this issue. If you have concerns that aren't covered by the link above, please open a separate issue to discuss them.
